base/strings/io.jl

```diff
@@ -123,6 +123,10 @@ function _str_sizehint(x)
         return sizeof(x)
     elseif x isa Char
         return ncodeunits(x)
+    elseif x isa UInt || x isa UInt32
```

Comment:
Could this just be `x isa Base.BitInteger` to catch all the cases?

Reply:
I don't think that's worth it - `UInt` and friends don't need the `x < 0` check and are faster without it (I haven't checked, but I think it's because the type determines the branch absolutely, so all other branches get compiled away. If I mix `Int` and `UInt` definitions, that doesn't happen an `UInt` becomes slower). `[U]Int16` has a maximum of 6 characters, which seems odd to grow an array by so feels fine to let it hit the `return 8`.

Reply:
> don't need the x < 0 check and are faster without it 

I am surprised about that, for example

```
julia> f(x) = x < 0;

julia> @code_llvm f(UInt(3))
;  @ REPL[35]:1 within `f'
define i8 @julia_f_1418(i64 zeroext %0) {
top:
  ret i8 0
}
```

Also if this was not the case, a single assembly instruction should be insignificant compared to e.g. the `ndigits` call, or?

Reply:
I do agree that it should not matter, all I'm saying is that I observed a difference 🤷‍♂️ Would be useful if you also ran the benchmark with and without this patch:

```patch
diff --git a/base/strings/io.jl b/base/strings/io.jl
index 7a940e32fc..de09e6ea40 100644                 
--- a/base/strings/io.jl                            
+++ b/base/strings/io.jl                            
@@ -123,9 +123,7 @@ function _str_sizehint(x)       
         return sizeof(x)                           
     elseif x isa Char                              
         return ncodeunits(x)                       
-    elseif x isa UInt || x isa UInt32              
-        return ndigits(x)                          
-    elseif x isa Int || x isa Int32                
+    elseif x isa BitInteger                        
         return ndigits(x) + (x < zero(x))          
     else                                           
         return 8                                   
```

It's kinda hard to benchmark this for me, since the overhead keeps changing and I can't seem to get to the fast 350-400ns space I was in earlier... code layout makes a mess of things, I think.

Reply:
New day, new try.

With explicit cases for UInt64/Int64:

```
julia> @benchmark string("x", "[", x, ",", y, "]") setup=(x=rand(UInt);y=rand(UInt)) evals=1
BenchmarkTools.Trial:                                                                       
  memory estimate:  432 bytes                                                               
  allocs estimate:  9                                                                       
  --------------                                                                            
  minimum time:     2.400 μs (0.00% GC)                                                     
  median time:      2.500 μs (0.00% GC)                                                     
  mean time:        2.687 μs (0.00% GC)                                                     
  maximum time:     19.500 μs (0.00% GC)                                                    
  --------------                                                                            
  samples:          10000                                                                   
  evals/sample:     1                                                                       
```

With BitInteger:

```
julia> @benchmark string("x", "[", x, ",", y, "]") setup=(x=rand(UInt);y=rand(UInt)) evals=1
BenchmarkTools.Trial:                                                                       
  memory estimate:  432 bytes                                                               
  allocs estimate:  9                                                                       
  --------------                                                                            
  minimum time:     3.900 μs (0.00% GC)                                                     
  median time:      4.100 μs (0.00% GC)                                                     
  mean time:        4.332 μs (0.00% GC)                                                     
  maximum time:     30.900 μs (0.00% GC)                                                    
  --------------                                                                            
  samples:          10000                                                                   
  evals/sample:     1                                                                                                                              
```

This is consistent across runs. 

There's one thing I can't explain though - the number of allocated bytes sometimes jumps to 448. Some alignment thing somewhere?

Reply:
I don't know how to investigate the difference further, I can only assume it's because `BitInteger` is a somewhat large `Union` - in which case, should `UInt128/Int128` be special cased as well? They can use up to 39 character to print after all, much more than the fallback 8.

Reply:
I don't know either why it would happen. I know this code is quite sensitive to "inference cycles" where if this code calls out into other code that ends up using this function (so you get a cycle), bad things can happen. Maybe it is something along those lines..? Just guessing.

Reply:
Hm, I haven't changed anything in terms of code that ran 🤔 I made sure to restart julia and only ran `using Revise, BenchmarkTools; Revise.track(Base); Revise.revise()` before each test with each version.

FWIW, if I don't split them and go with `BitInteger`, it's a regression - here's the same benchmark on my machine for the original code:

```
julia> using Revise, BenchmarkTools; Revise.track(Base); Revise.revise()                    
                                                                                            
julia> @benchmark string("x", "[", x, ",", y, "]") setup=(x=rand(UInt);y=rand(UInt)) evals=1
BenchmarkTools.Trial:                                                                       
  memory estimate:  480 bytes                                                               
  allocs estimate:  10                                                                      
  --------------                                                                            
  minimum time:     2.700 μs (0.00% GC)                                                     
  median time:      3.900 μs (0.00% GC)                                                     
  mean time:        4.580 μs (0.00% GC)                                                     
  maximum time:     77.700 μs (0.00% GC)                                                    
  --------------                                                                            
  samples:          10000                                                                   
  evals/sample:     1                                                                       
```

So I guess having the two cases split really is best, though I see no clear reson why - they at least both keep the allocations from growing.